### PR TITLE
win32: fit-border: keep the title bar inside the screen

### PR DESCRIPTION
**Pull request by avih**
```
Before this commit, when --fit-border was enabled (the default), the
title bar could end partially outside the screen in two cases:
- The window was auto-resized vertically to fit.
- Vertical resize was not needed to fit, but vertical-move was.

...
```
(see the commit message for full details).

The original code is d13f9d08 by @pavelxdd and @rossy - I'd appreciate if you could take a look at this commit.

### CODE REVIEW

**Comment by avih** [comment]
It might be worth noting that the original relevant comments are 100% correct, all of them along the lines of "adjust size and position if its size is larger". I.e. neither is adjusted if size is not larger (and no mention that the position adjustment might not be enough even when size if larger).

I think it could be due to few reasons:
- The issue couldn't/didn't happen at the time of the original code.
- It could happen but wasn't observed or not considered important to handle.
- It was observed but was too delicate to handle, and the comments tip-toed around it.

I suspect it was simply not observed, as the original commit made many changes at once, and this specific issue could easily gone unnoticed. However, that's only a gues.

File: video/out/w32_common.c [hunk 1]
```diff
@@ -795,33 +795,46 @@ static void update_window_style(struct vo_w32_state *w32)
...
 
-    // Calculate new position and save the rect
-    const int x = rc->left + o_w / 2 - n_w / 2;
-    const int y = rc->top + o_h / 2 - n_h / 2;
+    // Calculate new position with same center as before
+    int x = rc->left + o_w / 2 - n_w / 2;
+    int y = rc->top + o_h / 2 - n_h / 2;
+    // If we're not entirely inside rc2 - move just enough to make it inside
+    x = x < rc2->left ? rc2->left : x + n_w > rc2->right ? rc2->right - n_w : x;
+    y = y < rc2->top ? rc2->top : y + n_h > rc2->bottom ? rc2->bottom - n_h : y;
```
Comment (pavelxdd) [inline video/out/w32_common.c:832]: LGTM, except these lines are really hard to follow.

The original code didn't take into account the difference in rect positions, indeed.

Comment (avih) [reply video/out/w32_common.c:832]: Thanks, also for clarifying. I'll try to refactor these lines.

My main concern, however, is the big comment below with FIXME. Is it reasonably safe to assume that the fit code is not reached after fullscreen/maximized/minimized/sleep/etc?

It was too hard for me to follow the states, but I'll take your word if you feel it's reasonably safe (I'd prefer to understand though, if you care to explain).

File: video/out/w32_common.c [hunk 2]
```diff
@@ -974,7 +991,19 @@ static void reinit_window_state(struct vo_w32_state *w32)
...
+        // --fit-border is used) is contained entirely inside the screen.
+        // Therefore, it's important that it's not called when the user expects
+        // to restore a size/pos which was set manually, even if it's bigger
+        // and/or partially outside the screen, which is applicable at least
+        // when restoring from fullscreen/maximized/minimized/sleep/etc.
+        // This does seem to work correctly - empirically, however, the
+        // conditions which prevent calling it in these cases are delicate and
+        // hard to follow, with various state flags.
+        // We need to make it more obvious/documented how this call is avoided
+        // on such cases (i.e. how the next line is not reached).
```
Comment (pavelxdd) [inline video/out/w32_common.c:1004]: So, I assume from this comment, do you need to adjust the if statement above to skip calling the `fit_window_on_screen` in some more cases? I don't quite understand the problem here to be honest.

Comment (pavelxdd) [reply video/out/w32_common.c:1004]: `fit_window_on_screen` should not be called, unless there was a change in fullscreen state, or it was specifically asked to refit the window rect by setting the `w32->fit_on_screen`.

The `w32->fit_on_screen` is set to true when there's a change made to the window rect, so it needs to be refit on screen on next update.

File: video/out/w32_common.c (hunk 1 above)
Comment (pavelxdd) [reply video/out/w32_common.c:832]: I replied in the comment below about the conditions when `fit_window_on_screen` is called.

File: video/out/w32_common.c (hunk 2 above)
Comment (avih) [reply video/out/w32_common.c:1004]: The scenario which we don't want to reach is that the user manually moves/resizes the window to a state which fit would change it if called.

For instance, the user moves the window such that part of it is off-screen. Then they enter fullscreen or maximize or minimize, and then restore the window.

We want to restore it to the same position - partially off screen, but if the fit function would be called, then it will re-position the window inside the screen - which is not what the user expects.

In my attempts to trigger this case I couldn't make it happen, but I also couldn't guarantee that it will never happen, because the states at the code were too complex for me to follow.

So I'm asking whether it's safe to assume that we will not move the window into the screen while the user expects it to be restored to its original position.

Now is different than before because we now move the window if it's partially off screen, while previously fit only had an effect when the window size was too big.

Comment (pavelxdd) [reply video/out/w32_common.c:1004]: You could try running mpv with `geometry` option set, and see what will happen when you leave fullscreen.

Or you can try to trigger the conditions when `VOCTRL_SET_UNFS_WINDOW_SIZE` request is sent.

As far as I can see there are only 4 cases when `reinit_window_state` is called:
1) When restoring the maximized window in fullscreen state. (the `fit_window_on_screen` do nothing for maximized window anyway)
2) When there's a fullscreen state change detected (`w32->fit_on_screen` is not set in this case anyway).
3) In `gui_thread_reconfig` which is called on vo->reconfigure I suppose.
4) When the window size is changed via `VOCTRL_SET_UNFS_WINDOW_SIZE` request. This is the only case among 4 when `w32->fit_on_screen` flag could make an effect.

Comment (avih) [reply video/out/w32_common.c:1004]: > You could try running mpv with geometry ...
> Or you can try to trigger ...
> As far as I can see there are only 4 cases when reinit_window_state is called ...

Yeah, I know where it's called from, and I tried to trigger it in many ways, including multi-monitor setup in both horizontal and vertical configuration, and various other cases.

But ultimately that's anecdotal, because it's without deep understanding of how the states are updated, which WM_* handlers are called in which order, etc.

The point of the question was to understand whether you, as the original author, know, or have a good assessment of whether or not it could hit undesirable invocations.

**Comment by avih** [comment]
> So I'm asking whether it's safe to assume that we will not move the window into the screen while the user expects it to be restored to its original position.

OK, so after some sleep, I think I was asking the wrong question.

The problem I'm having is that this fit-border logic should only be applied on video-reconfig, together with geometry and other autofit options. However, its logic is at a function which is called from several different places, including `SC_RESTORE`, and on every fullscreen change.

The fit-border logic is simply at the wrong place.

It should only be called from `gui_thread_reconfig`, after `vo_calc_window_geometry2` and `vo_apply_window_geometry` (which do the geometry/autofit).

Or maybe I'm missing something some value it adds when called from SC_RESTORE and after fullscreen changes?

**Comment by pavelxdd** [comment]
SC_RESTORE or fullscreen changes should not trigger `fit_window_on_screen`, since `w32->fit_on_screen` is not set there, and there's a check for `toggle_fs` flag.

The only places it is triggered from are `gui_thread_reconfig` and `VOCTRL_SET_UNFS_WINDOW_SIZE` request.

Even if SC_RESTORE would somehow trigger `fit_window_on_screen`, there's a check for `IsMaximized(w32->window)` in this function, so it is safe to call it from there.

**Comment by avih** [comment]
> The only places it is triggered from are `gui_thread_reconfig` and `VOCTRL_SET_UNFS_WINDOW_SIZE` request.

Right. So this leaves the question of whether or not fit-border should be applied after `VOCTRL_SET_UNFS_WINDOW_SIZE`.

This VOCTRL is sent when the `window-scale` property changes. Considering that it doesn't apply the autofit logic (rightly, because autofit is only applied when we have new size for the video itself) then it should also not apply for fit-border.

So that's one place we now know it's a bug to apply it. I.e. we already know the patch in this PR is buggy here.

**Comment by avih** [comment]
@pavelxdd greatly appreciate your feedback. I'll think how to restructure it better and update the PR when I have something new.

Also, I feel better now that I can actually understand these states a bit better ;)

**Comment by avih** [comment]
@pavelxdd I'd appreciate if you could review these commits, preferably one by one.

I believe the only one which is potentially controversial is the last one (removal of `fit_on_screen` state), but after suffering while trying to understand what's going on with these states, I think I'm in a good position to clean it up with reasonable confidence, and hopefully with your approval.

Thanks.

File: video/out/w32_common.c [hunk 3]
```diff
@@ -831,7 +831,19 @@ static void fit_window_on_screen(struct vo_w32_state *w32)
...
+
+    if (w32->windowrc.top < screen.top) {
+        // if the top-edge of client area is above the target area (mainly
+        // because the client-area is centered but the title bar is taller
+        // than the bottom border), then move it down to align the edges.
+        // Windows itself applies the same constraint during manual move.
+        w32->windowrc.bottom += screen.top - w32->windowrc.top;
+        w32->windowrc.top = screen.top;
+        adjusted = true;
+    }
```
Comment (pavelxdd) [inline video/out/w32_common.c:838]: This logic can be moved inside `fit_rect` function, I think.

**Comment by pavelxdd** [comment]
I have no way to test these changes, but they look good to me.

I also think the last two commits can be squashed together, since they primarily work around the same code.

File: video/out/w32_common.c (hunk 3 above)
Comment (avih) [reply video/out/w32_common.c:838]: Technically yes, but logically I preferred them separated because one of them is technical rect handling (size) while this is a windows constraint.

**Comment by avih** [comment]
> I have no way to test these changes, but they look good to me.

I don't have a way to strictly test them either, because the logic of `!fs_toggle` combined with the `fit_on_screen` state and the other logic is just mind boggling to follow, but as the last commit suggests at the commit message, it's simply unnecessary.

> I also think the last two commits can be squashed together, since they primarily work around the same code.

Sure, it was mainly to make the last commit more local, only so that it's easier to understand what it does. Clearly a no-op commit can be combined, though I still think it has some value both on its own and in making the following commit slightly easier to follow.

If you insist on squashing them then sure, I could do that.

Same for moving the too-high logic into fit_rect, if you insist, I'll do that.

**Comment by avih** [comment]
> I have no power

That's why I gave you the power. Anyway, it doesn't sound like you insist, so I'll assume you can live with these.

However, I'm now looking again at your refactor commit, and I think I now know what `fit_on_screen` was really meant fot.

It was meant to fix your own issue - #5215 - in multi-monitor config, when manually resizing to a size too big for one monitor (but still valid in multi monitor), and then going into and out of fullscreen - the size was fit incorrectly to one monitor size.

And I'm guessing the whole refactor of d13f9d0 was created as a result.

So basically, `fit_on_screen` is set whenever a new size comes from mpv (init, window-scale, etc), and not set elsewhere:
- During manual user interaction with the window (resize, restore, etc)
- When restoring to a size which the user set previously manually.
- When only a style change is needed (ontop, border - doesn't border need fit though in case the borderless size is ok but bordered is not?).

And indeed, the use case of #5215 is broken with the last commit, and the commit message is incorrect by saying "fit_window_on_screen is 100% harmless ... it rightly doesn't limit the size to one monitor-size ".

That's because the size is for one monitor, which also makes it harmful when restoring to a manually-set size which is valid in multi-mon but too big for one minitor.

I think I'll just drop the last two commits from this PR.

While I am in a good position to clean up the states mess, I really don't feel like spending more time on these states than I already have.

**Comment by avih** [comment]
@pavelxdd I've updated only the last commit and made it more conservative, but still simpler.

I'd appreciate your opinion on it.

Additionally, there's one call to `reinit_window_state` which I simply cannot reach. The original code was added by you at 3739d131 (`if (IsMaximized(w32->window) && w32->current_fs)` at the `case SC_RESTORE:` clause) - I just can't reach a state where we're both maximized and current_fs is true.

The commit message mentions "maximized window restored from fullscreen", but if I maximize it, and then go fullscreen, I don't know how to restore it. The window menu (alt-space) is active, but the "Restore" menu is grayed out. Also, if I toggle fullscreen to get back to maximized, it works but doesn't go through this code.

How can I reach this code?

**Comment by pavelxdd** [comment]
Maybe you can try this: 
1) maximize the window
2) go to fullscreen
3) press Win+D to minimize it
4) click on mpv window on taskbar to restore it

The last commit seems good to me.

**Comment by avih** [comment]
Thanks, though this procedure didn't work for me to enter the if clause. Well.. whatever.

Thanks for the reviews. I'll push it as is, and we'll handle issues if they come up.

**Comment by avih** [comment]
Pushed to master, 4 commits ending at f60e327e49e5ccf40a4988716be7ccae0882d36e. Thanks again :)
